File: Train/cpp/1_TwoSum/1_TwoSum.cpp
#include <unordered_map>
#include <vector>
#include <gtest/gtest.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mem;  // value to index
        for (int i = 0; i < nums.size(); ++i) {
            if (mem.count(target - nums[i]) != 0)
                return {i, mem[target - nums[i]]};
            mem[nums[i]] = i;
        }
        return {-1, -1};
    }
};

TEST(test, case0) {
    vector<int> input({2,7,11,15});
    int target = 9;
    vector<int> output1{1, 0};
    vector<int> output2{0, 1};
    Solution s;
    EXPECT_TRUE(output1 == s.twoSum(input, target) || output2 == s.twoSum(input, target));
}

File: Train/go/1_TwoSum/1_TwoSum.go
package TwoSum
func twoSum(nums []int, target int) []int {
    mem := make(map[int]int) // value to index
    for i, value := range nums {
        if index, ok := mem[target - value]; ok {
            return []int{i, index}
        }
        mem[nums[i]] = i
    }
    return []int{-1, -1}
}

